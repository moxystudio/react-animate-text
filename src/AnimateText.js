import React, { useCallback, useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import SplitText from '@moxy/react-split-text';
import IntersectionObserver from '@researchgate/react-intersection-observer';
import 'intersection-observer'; // Polyfill to support older browsers

const AnimateText = ({
    children,
    className,
    separator,
    visible,
    transitionDuration,
    initialDelay,
    lineDelay,
    wordDelay,
}) => {
    const containerRef = useRef(null);
    const [isReady, setIsReady] = useState(false);
    const [shouldAnimate, setShouldAnimate] = useState(visible ?? false);

    useEffect(() => {
        const elements = [...containerRef.current.children];
        let prevOffset = 0;
        let delay = initialDelay;

        elements.forEach((word, index) => {
            const newLine = word.offsetTop > prevOffset;

            if (newLine) {
                delay += lineDelay;
            } else if (wordDelay && index > 0) {
                delay += wordDelay;
            }

            word.style.transitionDelay = `${delay}s`;

            prevOffset = word.offsetTop;
        });

        setIsReady(true);
    }, [initialDelay, lineDelay, wordDelay]);

    useEffect(() => {
        setShouldAnimate(visible);
    }, [visible]);

    const handleIntersection = useCallback(({ isIntersecting }) => {
        isIntersecting && setShouldAnimate(true);
    }, []);

    const renderAnimateText = () => (
        <div
            ref={ containerRef }
            data-testid="react-animate-text"
            aria-hidden={ !(isReady && shouldAnimate) }
            className={ classNames('react-animate-text_container', className) }
            style={ { transitionDuration: `${transitionDuration}s` } }>
            <SplitText separator={ separator } data-attribute="word" style={ { transitionDuration: `${transitionDuration}s` } }>
                { children }
            </SplitText>
        </div>
    );

    return typeof visible === 'boolean' ?
        renderAnimateText() :
        <IntersectionObserver onChange={ handleIntersection }>
            { renderAnimateText() }
        </IntersectionObserver>;
};

AnimateText.propTypes = {
    children: PropTypes.string.isRequired,
    className: PropTypes.string,
    separator: PropTypes.string,
    visible: PropTypes.bool,
    transitionDuration: PropTypes.number,
    initialDelay: PropTypes.number,
    lineDelay: PropTypes.number,
    wordDelay: PropTypes.number,
};

AnimateText.defaultProps = {
    separator: ' ',
    transitionDuration: 0.8,
    initialDelay: 0,
    lineDelay: 0.3,
    wordDelay: 0,
};

export default AnimateText;
