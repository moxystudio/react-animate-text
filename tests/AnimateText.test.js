import React from 'react';
import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import AnimateText from '../src/AnimateText';

let onChangeObserver;

jest.mock('@researchgate/react-intersection-observer', () => ({ children, onChange }) => {
    onChangeObserver = onChange;

    return children;
});

const renderWithProps = (props = {}) => render(
    <AnimateText { ...props }>
        One two three four five
    </AnimateText>,
);

describe('AnimateText with a single line', () => {
    it('should render correctly', () => {
        const { getByText } = renderWithProps();

        expect(getByText(/one/i)).toBeInTheDocument();
        expect(getByText(/five/i)).toBeInTheDocument();
    });

    it('should render with className', () => {
        const { getByTestId } = renderWithProps({ className: 'container-class' });

        expect(getByTestId('react-animate-text')).toHaveClass('container-class');
    });

    it('should have a transition duration of 2s', () => {
        const { getByTestId } = renderWithProps({ transitionDuration: 2 });

        expect(getByTestId('react-animate-text')).toHaveStyle({ transitionDuration: '2s' });
    });

    it('should have an initial delay of 1s', () => {
        const { getByText } = renderWithProps({ initialDelay: 1 });

        expect(getByText(/one/i)).toHaveStyle({ transitionDelay: '1s' });
    });

    it('should have an word animation delay of 2s', () => {
        const { getByText } = renderWithProps({ wordDelay: 2, initialDelay: 0 });

        expect(getByText(/one/i)).toHaveStyle({ transitionDelay: '0s' });
        expect(getByText(/two/i)).toHaveStyle({ transitionDelay: '2s' });
        expect(getByText(/three/i)).toHaveStyle({ transitionDelay: '4s' });
        expect(getByText(/four/i)).toHaveStyle({ transitionDelay: '6s' });
        expect(getByText(/five/i)).toHaveStyle({ transitionDelay: '8s' });
    });

    it('should set the aria-hidden attribute correctly', () => {
        const { getByTestId } = renderWithProps();

        act(() => {
            onChangeObserver({ isIntersecting: true });
        });

        expect(getByTestId('react-animate-text')).toHaveAttribute('aria-hidden', 'false');

        act(() => {
            onChangeObserver({ isIntersecting: false });
        });

        expect(getByTestId('react-animate-text')).toHaveAttribute('aria-hidden', 'false');
    });

    it('should render correctly with show prop', () => {
        const { getByTestId, rerender } = renderWithProps({ show: false });

        expect(getByTestId('react-animate-text')).toHaveAttribute('aria-hidden', 'true');

        rerender(
            <AnimateText show>
                Example Text
            </AnimateText>,
        );

        expect(getByTestId('react-animate-text')).toHaveAttribute('aria-hidden', 'false');
    });

    it('should not show the text', () => {
        const { getByTestId } = renderWithProps({ shouldShow: false });

        expect(getByTestId('react-animate-text')).toHaveAttribute('aria-hidden', 'true');
    });
});

describe('AnimateText with multiple lines', () => {
    beforeAll(() => {
        Object.defineProperty(HTMLElement.prototype, 'offsetTop', {
            get() {
                const words = Array.prototype.slice.call(this.parentElement.children);
                const index = words.indexOf(this);

                // First line
                if (index < 2) {
                    return 0;
                }

                // Third line
                if (index > 2) {
                    return 20;
                }

                // Second line
                return 10;
            },
        });
    });

    it('should render correctly', () => {
        const { asFragment } = renderWithProps();

        expect(asFragment()).toMatchSnapshot();
    });

    it('should use the line delay correctly', () => {
        const { asFragment } = renderWithProps({ lineDelay: 1 });

        expect(asFragment()).toMatchSnapshot();
    });

    it('should use the line delay correctly along with word delay', () => {
        const { asFragment } = renderWithProps({ lineDelay: 1, wordDelay: 0.5 });

        expect(asFragment()).toMatchSnapshot();
    });
});
