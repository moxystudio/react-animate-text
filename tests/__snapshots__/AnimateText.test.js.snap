// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimateText with multiple lines should render correctly 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="react-animate-text_container"
    data-testid="react-animate-text"
    style="transition-duration: 0.8s;"
  >
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 0s;"
    >
      One 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 0s;"
    >
      two 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 0.3s;"
    >
      three 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 0.6s;"
    >
      four 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 0.6s;"
    >
      five
    </span>
  </div>
</DocumentFragment>
`;

exports[`AnimateText with multiple lines should use the line delay correctly 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="react-animate-text_container"
    data-testid="react-animate-text"
    style="transition-duration: 0.8s;"
  >
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 0s;"
    >
      One 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 0s;"
    >
      two 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 1s;"
    >
      three 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 2s;"
    >
      four 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 2s;"
    >
      five
    </span>
  </div>
</DocumentFragment>
`;

exports[`AnimateText with multiple lines should use the line delay correctly along with word delay 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="react-animate-text_container"
    data-testid="react-animate-text"
    style="transition-duration: 0.8s;"
  >
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 0s;"
    >
      One 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 0.5s;"
    >
      two 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 1.5s;"
    >
      three 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 2.5s;"
    >
      four 
    </span>
    <span
      class="react-split-text"
      data-attribute="word"
      style="transition-duration: 0.8s; transition-delay: 3s;"
    >
      five
    </span>
  </div>
</DocumentFragment>
`;
